src/py_adaptors.h

```diff
@@ -21,6 +22,84 @@ int convert_path(PyObject *obj, void *pathp);
 namespace py
 {
 
+template<typename T, size_t Dim>
+class StridedMemoryBase {
+protected:
+    char *m_data;
+    npy_intp *m_strides;
+
+    explicit StridedMemoryBase(char *data, npy_intp *strides)
+        : m_data(data), m_strides(strides)
+    {
+    }
+
+public:
+    StridedMemoryBase()
+        : m_data(nullptr), m_strides(nullptr)
+    {
+    }
+
+    explicit StridedMemoryBase(PyArrayObject *array)
+    {
+        if (PyArray_NDIM(array) != Dim) {
+            PyErr_SetString(PyExc_ValueError, "Invalid array dimensionality");
+            throw py::exception();
+        }
+        m_data = PyArray_BYTES(array);
+        m_strides = PyArray_STRIDES(array);
+    }
+
+    operator bool() const {
+        return m_data != nullptr;
+    }
+
+    void reset() {
+        m_data = nullptr;
+        m_strides = nullptr;
+    }
+
+    T* data() {
+        return reinterpret_cast<T*>(m_data);
+    }
+};
+
+#define INHERIT_CONSTRUCTORS(CLS, DIM)                                         \
+  CLS() {}                                                                     \
+  explicit CLS(char *data, npy_intp *strides): StridedMemoryBase<T, DIM>(data, strides) {} \
+  explicit CLS(PyArrayObject *array): StridedMemoryBase<T, DIM>(array) {}
```

Comment:
This is a bit awkward, but I don't think there's a way around it unless I can use C++11. Are there plans to migrate to C++11 by any chance?